File: Number of Provinces - GFG/number-of-provinces.cpp
// { Driver Code Starts
#include <bits/stdc++.h>
using namespace std;


 // } Driver Code Ends
//User function Template for C++

class Solution {
  public:
    int numProvinces(vector<vector<int>> adj, int V) {
        // code here
        return solve(adj, V);
    }
  private:
    void dfs(vector<vector<int>>& adj, vector<bool>& visited, int v, int node) {
        visited[node] = true;
        
        for(int nbr = 0; nbr < v; ++nbr) {
            if(adj[node][nbr] == 1 && visited[nbr] == false)
                dfs(adj, visited, v, nbr);
        }
    }
  
    int solve(vector<vector<int>>& adj, int v) {
        vector<bool> visited(v, false);
        int ans = 0;
        
        for(int i = 0; i < v; ++i) {
            if(visited[i] == false) {
                dfs(adj, visited, v, i);
                ++ans;
            }
        }
        
        return ans;
    }
 
};

// { Driver Code Starts.

int main() {
    int t;
    cin >> t;
    while (t--) {
        int V,x;
        cin>>V;
        
        vector<vector<int>> adj;
        
        for(int i=0; i<V; i++)
        {
            vector<int> temp;
            for(int j=0; j<V; j++)
            {
                cin>>x;
                temp.push_back(x);
            }
            adj.push_back(temp);
        }
        

        Solution ob;
        cout << ob.numProvinces(adj,V) << endl;
    }
    return 0;
}  // } Driver Code Ends

File: Number of Provinces - GFG/number-of-provinces.py
#User function Template for python3

class Solution:
    
    def dfs(self, node, visited, adj, v):
        visited[node] = True
        
        for i in range(v):
            if visited[i] == False and adj[node][i]:
                self.dfs(i, visited, adj, v)

    
    def numProvinces(self, adj, v):
        visited = [False] * v
        ans = 0
        for i in range(v):
            if visited[i] == False:
                self.dfs(i, visited, adj, v)
                ans += 1
        return ans
        
        

#{ 
#  Driver Code Starts
#Initial Template for Python 3

if __name__ == '__main__': 
    t = int (input ())
    for _ in range (t):
        V=int(input())
        adj=[]
        
        for i in range(V):
            temp = list(map(int,input().split()))
            adj.append(temp);
        
        ob = Solution()
        print(ob.numProvinces(adj,V))
# } Driver Code Ends